### PR TITLE
Loki: add from JSON to dataframes tests

### PR DESCRIPTION
**Pull request by gabor**
(NOTE: this is the same for Loki as https://github.com/grafana/grafana/blob/main/pkg/tsdb/prometheus/framing_test.go is for Prometheus)

there is a small change in `pkg/tsdb/loki/loki.go`, we extract a part of  the `QueryData` function into a separate function, to be able to write tests for it.
the rest of the code is new code, the test code and "input" and "output" snapshot files.

we add tests that test the whole dataframe processing, starting with the Loki JSON response, and seeing how it gets transformed into grafana dataframes.

notes:
1. the tests test what is the current behavior in the Loki alerting code. this may not be what is "correct". (for example, the error-message returned when an error happens is not good)
1. i am testing error-responses too, and those are different in different Loki versions. some versions return it as plaintext, just a text message like "this is wrong", and some other Loki versions return it as a JSON structure. also, some versions return a JSON structure, but use the `text/plain` content type. so i am testing all 3 combinations.

how to test:
- ( this is mostly about adding unit-tests, which are verified by the CI )
- to verify that the backend-changes did not break alerting in Loki, try a simple alerting scenario in Loki, and if it works, it should be ok

### CODE REVIEW

File: pkg/tsdb/loki/framing_test.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+		{
+			name:         "parse a non-json error body with json content type (loki 2.4.0,2.4.1,2.4.2)",
+			body:         []byte("parse error at line 1, col 8: something is wrong"),
+			contentType:  "application/json; charset=UTF-8",
+			errorMessage: "Run out of attempts while querying the server",
+		},
+		{
+			name:         "parse an error response in plain text",
+			body:         []byte("parse error at line 1, col 8: something is wrong"),
+			contentType:  "application/json; charset=utf-8",
```
Comment (ivanahuckova) [inline pkg/tsdb/loki/framing_test.go:97]: Should this be 
```suggestion
			contentType:  "text/plain; charset=UTF-8",
```
?

**Review by ivanahuckova — COMMENTED** [review]
I've left 1 question and 1 suggestion below. Looks good!

File: pkg/tsdb/loki/testdata/matrix_inf.json [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    "result": [
+      {
+        "metric": {
+          "level": "info",
+          "location": "moon",
+          "protocol": "http"
+        },
+        "values": [
+          [1643021611.9, "+Inf"],
+          [1643021911.9, "+Inf"]
```
[the remark this answers is not in the document]
Comment (gabor) [reply pkg/tsdb/loki/testdata/matrix_inf.json]: good idea, done!

File: pkg/tsdb/loki/framing_test.go (hunk 1 above)
Comment (gabor) [reply pkg/tsdb/loki/framing_test.go]: good catch, i got mixed up in those content types 😄 . fixed.

**Review by ivanahuckova — APPROVED** [review]
👍

**Pull request merged**